File: 547. Number of Provinces/Solution.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        uf = UFDS(n)
        for x in range(n):
            for y in range(n):
                if isConnected[x][y] == 1: uf.union(x, y)
        return uf.n

class UFDS:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank = [0 for i in range(n)]
        self.n = n
    
    def find(self, x):
        if self.parent[x] != x: self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def same(self, x, y):
        return self.find(x) == self.find(y)
    
    def union(self, x, y):
        if self.same(x, y): return
        a, b = self.find(x), self.find(y)
        if self.rank[a] > self.rank[b]:
            self.parent[b] = a
        else:
            self.parent[a] = b
            if self.rank[a] == self.rank[b]: self.rank[b] += 1
        self.n -= 1

File: 547. Number of Provinces/Solution.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        UFDS uf = new UFDS(n);
        for (int x = 0; x < n; x++) {
            for (int y = 0; y < n; y++) {
                if (isConnected[x][y] == 1) uf.union(x, y);
            }
        }
        return uf.N;
    }
}

class UFDS {
    int[] parent, rank;
    int N;

    public UFDS(int n) {
        parent = new int[n];
        rank = new int[n];
        N = n;
        for (int i = 0; i < n; i++) parent[i] = i;
    }

    public int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    public boolean same(int x, int y) {
        return find(x) == find(y);
    }

    public void union(int x, int y) {
        if (same(x, y)) return;
        int a = find(x), b = find(y);
        if (rank[a] > rank[b]) {
            parent[b] = a;
        } else {
            parent[a] = b;
            if (rank[a] == rank[b]) rank[b]++;
        }
        N--;
    }
}
